### PR TITLE
[SR GeekCamp]merge join fe （issue:4036）

### LINKED ISSUE #4036: Optimizer supports Merge Join【Challenging】
# Requirement Description
In some specific scenarios, such as all tables are naturally ordered, non-equivalent Join, or other possible scenarios, MergeJoin is better than HashJoin in performance and memory usage. It is necessary to explore possible scenarios and modify the optimizer, Properly plan MergeJoin based on statistics or metadata information.

# Mentor
@trueeyu 

# Registration procedure

**Step 1.** Add a comment when you sign up. The comment can be a brief self-introduction and reason for signing up, including what you do, what you're good at, what you're interested in and your technology stack, and how familiar you are with StarRocks. Example:

> I pick! My name is Johnny. I am from StarRocks and I have been working on database kernel development for 1 year. I am interested in OLAP, good at C++ programming, and have been using StarRocks for 6 months. I chose to solve this issue because I am familiar with execution engine.

**Step 2.** On your WeChat, add the StarRocks community assistant "**_StarRocks-1_**" as a WeChat friend and leave your GitHub ID. The assistant will invite you to a WeChat group based on the issue you picked.

**Step 3.** The mentor will communicate with all the participants who picked this issue and assign the task to the best match. 

# Before you start
1. Read the StarRocks [GitHub workflow](https://github.com/StarRocks/community/blob/main/Contributors/guide/workflow.md) carefully, which provides the configurations of BE and FE and describes how to submit a pull request.
2. To learn more about StarRocks, read [StarRocks documentation](https://github.com/StarRocks/docs/blob/main/TOC.md).

# Pull request requirements

1. After you complete code development, upload the code to the StarRocks code repo as a pull request (PR).

  - PR title format: **[SR GeekCamp] + PR title**
  - Labels ：[Geek Camp 2022/feature]
2. After you submit your PR, comment on the status of your PR submission as requested under the Pinned Issue [[StarRocks geek Camp 2022] ](https://github.com/StarRocks/starrocks/issues/4008)page.

  - Comment format: **"I have finished Issue #" + PR link id**

### PR DESCRIPTION
**Pull request by zddr**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4036

### CODE REVIEW

**Comment by mofeiatwork** [comment]
What's your plan for the merge join algorithm?

**Comment by zddr** [comment]
> What's your plan for the merge join algorithm?

The general idea is to judge whether it is ordered by default. If it is disordered, the sorting cost will be added when calculating the cost. Logically, the cost will be bigger than that of hash join. If it is ordered, the sorting cost will not be calculated, and it will be less than that of hash join. Of course, if there are other situations suitable for merge join in the future, it can also continue to be in costmodel.Visitphysicalmergejoin extension.

**Comment by zddr** [comment]
At present, I mainly want to determine the main idea and whether the framework transformation is reasonable, and then I will focus on the merge join algorithm.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 1]
```diff
@@ -345,6 +350,64 @@ public CostEstimate visitPhysicalHashJoin(PhysicalHashJoinOperator join, Express
...
+                    return CostEstimate.of((leftStatistics.getOutputSize(context.getChildOutputColumns(0))
+                                    + rightStatistics.getOutputSize(context.getChildOutputColumns(1)) / 2),
+                            0, 0);
+                } else {
+                    return CostEstimate.of((leftStatistics.getOutputSize(context.getChildOutputColumns(0))
+                                    + rightStatistics.getOutputSize(context.getChildOutputColumns(1))) * 1.5,
+                            leftStatistics.getOutputSize(context.getChildOutputColumns(0))
+                                    + rightStatistics.getOutputSize(context.getChildOutputColumns(1)), 0);
+                }
+            }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:386]: Why need checkSort? sort cost will compute in `PhysicalTopNNode`

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java [hunk 2]
```diff
@@ -495,7 +495,7 @@ private PhysicalHashAggregateOperator rewriteAggOperator(PhysicalHashAggregateOp
                     int columnId = kv.getValue().getUsedColumns().getFirstId();
                     if (context.needRewriteMultiCountDistinctColumns.contains(columnId)) {
                         // we only need rewrite TFunction
-                        Type[] newTypes = new Type[] {ID_TYPE};
+                        Type[] newTypes = new Type[]{ID_TYPE};
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java:498]:
```suggestion
                        Type[] newTypes = new Type[] {ID_TYPE};
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java [hunk 3]
```diff
@@ -318,4 +324,14 @@ public List<Rule> getImplementRules() {
...
+
+    public void removeImplementationRule(Class<? extends ImplementationRule> implementationRuleClass) {
+        Iterator<Rule> iterator = implementRules.iterator();
+        while (iterator.hasNext()) {
+            Rule rule = iterator.next();
+            if (rule.getClass() == implementationRuleClass) {
+                iterator.remove();
+            }
+        }
+    }
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/RuleSet.java:336]: Not a nice implement, your can refer by `transformRules`

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 4]
```diff
@@ -1794,6 +1797,294 @@ public PlanFragment visitPhysicalHashJoin(OptExpression optExpr, ExecPlan contex
             }
         }
 
+        @Override
+        public PlanFragment visitPhysicalMergeJoin(OptExpression optExpr, ExecPlan context) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1801]: What different with `HashJoin`? can be refactor together?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 266 / 732 (36.34%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|0|28|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMergeJoinOperator.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinPredicateUtils.java|0|16|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/planner/MergeJoinNode.java|0|50|00.00%|
|:large_blue_circle:|com/starrocks/sql/Explain.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpressionVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|1|17|05.88%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|22|234|09.40%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MergeJoinImplementationRule.java|2|12|16.67%|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|152|243|62.55%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJoinOperator.java|29|39|74.36%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|4|5|80.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|9|10|90.00%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/HashJoinImplementationRule.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashJoinOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/JoinImplementationRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|3|3|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 261 / 737 (35.41%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|0|28|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMergeJoinOperator.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinPredicateUtils.java|0|16|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MergeJoinImplementationRule.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/planner/MergeJoinNode.java|0|50|00.00%|
|:large_blue_circle:|com/starrocks/sql/Explain.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpressionVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|1|17|05.88%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|22|234|09.40%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|4|9|44.44%|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|152|243|62.55%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|8|12|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJoinOperator.java|29|39|74.36%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|9|10|90.00%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/HashJoinImplementationRule.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashJoinOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/JoinImplementationRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|3|3|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 5]
```diff
@@ -1334,13 +1337,13 @@ public void rewriteAggDistinctFirstStageFunction(List<FunctionCallExpr> aggregat
                 if (functionName.equalsIgnoreCase(FunctionSet.COUNT)) {
                     replaceExpr = new FunctionCallExpr(FunctionSet.MULTI_DISTINCT_COUNT, functionCallExpr.getParams());
                     replaceExpr.setFn(Expr.getBuiltinFunction(FunctionSet.MULTI_DISTINCT_COUNT,
-                            new Type[] {functionCallExpr.getChild(0).getType()},
+                            new Type[]{functionCallExpr.getChild(0).getType()},
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1340]:
```suggestion
                            new Type[] {functionCallExpr.getChild(0).getType()},
```

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 6]
```diff
@@ -1334,13 +1337,13 @@ public void rewriteAggDistinctFirstStageFunction(List<FunctionCallExpr> aggregat
...
                             IS_NONSTRICT_SUPERTYPE_OF));
                     replaceExpr.getParams().setIsDistinct(false);
                 } else if (functionName.equalsIgnoreCase("SUM")) {
                     replaceExpr = new FunctionCallExpr(FunctionSet.MULTI_DISTINCT_SUM, functionCallExpr.getParams());
                     replaceExpr.setFn(Expr.getBuiltinFunction(FunctionSet.MULTI_DISTINCT_SUM,
-                            new Type[] {functionCallExpr.getChild(0).getType()},
+                            new Type[]{functionCallExpr.getChild(0).getType()},
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1346]:
```suggestion
                            new Type[] {functionCallExpr.getChild(0).getType()},
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 264 / 732 (36.07%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|0|28|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMergeJoinOperator.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinPredicateUtils.java|0|16|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MergeJoinImplementationRule.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/planner/MergeJoinNode.java|0|50|00.00%|
|:large_blue_circle:|com/starrocks/sql/Explain.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpressionVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|1|17|05.88%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|20|229|08.73%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|4|9|44.44%|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|154|243|63.37%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|4|6|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJoinOperator.java|30|39|76.92%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|10|12|83.33%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|9|10|90.00%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/HashJoinImplementationRule.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashJoinOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/JoinImplementationRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|3|3|100.00%|

File: gensrc/thrift/PlanNodes.thrift [hunk 7]
```diff
@@ -405,6 +405,37 @@ struct THashJoinNode {
   54: optional list<Types.TSlotId> output_columns
 }
 
+struct TMergeJoinNode {
+  1: required TJoinOp join_op
```
Comment (trueeyu) [inline gensrc/thrift/PlanNodes.thrift:409]: all the items of Thtrift  better set to optional

File: fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java [hunk 8]
```diff
@@ -505,6 +506,9 @@ public class SessionVariable implements Serializable, Writable, Cloneable {
     @VariableMgr.VarAttr(name = ENABLE_HIVE_COLUMN_STATS)
     private boolean enableHiveColumnStats = true;
 
+    @VariableMgr.VarAttr(name = JOIN_IMPLEMENTATION_MODE)
+    private String joinImplementationMode = "hash"; // auto, merge, hash
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/qe/SessionVariable.java:510]: Should check the session variables  when set, must be auto, merge and hash

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 9]
```diff
@@ -131,6 +131,16 @@ void memoOptimize(ConnectContext connectContext, Memo memo, TaskContext rootTask
             }
         }
 
+        //add join implementRule
+        String joinImplementationMode = ConnectContext.get().getSessionVariable().getJoinImplementationMode();
+        if ("merge".equals(joinImplementationMode)) {
```
Comment (trueeyu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:136]: equalsIgnoreCase ？

Comment (zddr) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:136]: good idea

Comment (zddr) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java]: done

**Review by trueeyu — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 10]
```diff
@@ -1863,6 +1868,316 @@ public PlanFragment visitPhysicalHashJoin(OptExpression optExpr, ExecPlan contex
             }
         }
 
+        @Override
+        public PlanFragment visitPhysicalMergeJoin(OptExpression optExpr, ExecPlan context) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:1872]: This method is too repetitive. I think we should reuse the logic of hashjoin

File: fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java [hunk 11]
```diff
@@ -1877,6 +2192,68 @@ private boolean isColocateJoin(OptExpression optExpression) {
...
+            }
+            if (operator instanceof PhysicalDistributionOperator) {
+                return;
+            }
+            for (OptExpression child : optExpression.getInputs()) {
+                collectOlapScanInFragment(child, scanNodeList);
+            }
+        }
+
+        private boolean isColocateJoin(OptExpression optExpression, ExecPlan context, PlanNode left, PlanNode right) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/plan/PlanFragmentBuilder.java:2210]: Unused?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 402 / 577 (69.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/Explain.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/planner/MergeJoinNode.java|3|50|06.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalMergeJoinOperator.java|5|14|35.71%|
|:large_blue_circle:|com/starrocks/planner/HashJoinNode.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/planner/JoinNode.java|154|243|63.37%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|6|9|66.67%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalJoinOperator.java|30|39|76.92%|
|:large_blue_circle:|com/starrocks/planner/RuntimeFilterDescription.java|11|14|78.57%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|6|83.33%|
|:large_blue_circle:|com/starrocks/planner/PlanFragment.java|10|12|83.33%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/JoinPredicateUtils.java|14|16|87.50%|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|69|72|95.83%|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|27|28|96.43%|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/HashJoinImplementationRule.java|4|4|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/MergeJoinImplementationRule.java|14|14|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHashJoinOperator.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/JoinImplementationRule.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptExpressionVisitor.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/OperatorType.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/planner/PlanNode.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|17|17|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java|10|10|100.00%|

**Review by Seaven — APPROVED** [review]

**Review by trueeyu — APPROVED** [review]

**Pull request merged**